torch/distributions/constraints.py

```diff
@@ -0,0 +1,104 @@
+import torch
+from torch.nn.functional import sigmoid, softmax
+
+
+class Constraint(object):
+    """
+    Abstract base class for constraints.
+
+    A constraint object represents a region over which a continuous variable is
+    valid, e.g. within which a variable can optimized.
+
+    Each constraint class registers a pseudoinverse pair of transforms
+    `to_unconstrained` and `from_unconstrained`. These allow standard
+    parameters to be transformed to an unconstrained space for optimization and
+    transformed back after optimization. Note that these are not necessarily
+    inverse pairs since the unconstrained space may have extra dimensions that
+    are projected out; only the one-sided inverse equation is guaranteed::
+
+        x == c.to_constrained(c.to_unconstrained(x))
+    """
+    def to_unconstrained(self, x):
+        """
+        Transform from constrained coordinates to unconstrained coordinates.
+        """
+        raise NotImplementedError
+
+    def to_constrained(self, x):
+        """
+        Transform from unconstrained coordinates to constrained coordinates.
+        """
+        raise NotImplementedError
+
+
+class Unconstrained(Constraint):
+    """
+    Trivial constraint to the real line `(-inf, inf)`.
+    """
+    def to_unconstrained(self, x):
+        return x
+
+    def to_constrained(self, u):
+        return u
+
+
+class Dependent(Constraint):
+    """
+    Placeholder for variables whose support depends on other variables.
+    """
+    def to_unconstrained(self, x):
+        raise ValueError('Parameter cannot be transformed from an unconstrained space; '
+                         'Try another parameterization to make parameters independent.')
+
+    def to_constrained(self, u):
+        raise ValueError('Parameter cannot be transformed to an unconstrained space; '
+                         'Try another parameterization to make parameters independent.')
+
+
+class Positive(Constraint):
+    """
```

Comment:
This is nice! Do we see a need to implement multiple transform pairs for each constraint type in the future? If so, each constraint type could register multiple transform pairs (or register bijectors to transform samples), with a default pair like the one you already have. We can probably defer this until we really need it though.

Reply:
It seems like a reasonable pattern for extension is to subclass `Positive` with something like
```py
class PositiveWithSoftplus(Positive):
    def to_constrained(u): ...
    def to_unconstrained(x): ...
```
The other direction `Constraint` class will grow is to add `.bijector` properties which are different from the constraint transforms implemented in this PR. These will look something like
```py
class Positive(Constraint):
    def to_constrained(u): ...
    def to_unconstrained(x): ...
    bijector = [LogBijector()]

class Interval(Constraint):
    ...
    @property
    def bijector(self):
        return [AffineBijector(self.lower_bound, self.upper_bound - self.lower_bound),
                SigmoidBijector()]
```

Reply:
Nice, I like it.